src/kakarot/instructions/stop_and_arithmetic_operations.cairo

```diff
@@ -63,468 +44,374 @@ namespace StopAndArithmeticOperations {
+        // ADDMOD;
+        dw 0x08;
+        dw 8;
+        dw 3;
+        dw -2;
+        dw 0;
+        // MULMOD;
+        dw 0x09;
+        dw 8;
+        dw 3;
+        dw -2;
+        dw 0;
+        // EXP;
+        dw 0x0A;
+        dw 10;
+        dw 2;
+        dw -1;
+        dw 0;
+        // SIGNEXTEND;
+        dw 0x0B;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+    }
 
-        let stack_underflow = is_le(ctx.stack.size, 1);
+    func exec_arithmetic_operation{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*, opcode: model.Opcode*) -> model.ExecutionContext* {
+        alloc_locals;
+        local stack: model.Stack*;
+
+        let stack_underflow = is_le(ctx.stack.size, opcode.stack_input - 1);
         if (stack_underflow != 0) {
             let (revert_reason_len, revert_reason) = Errors.stackUnderflow();
             let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
             return ctx;
         }
 
-        // Stack input:
-        // 0 - a: number.
-        // 1 - b: exponent.
-        let stack = ctx.stack;
-        let (stack, popped) = Stack.pop_n(self=stack, n=2);
+        let (_stack, popped) = Stack.pop_n(ctx.stack, opcode.stack_input);
+
+        assert stack = _stack;
+        tempvar offset = 1 + 4 * (opcode.number - 0x1);
+
+        // Prepare arguments
+        [ap] = range_check_ptr, ap++;
+        [ap] = popped, ap++;
+
+        // call opcode
+        jmp rel offset;
+        call add;  // 0x1
+        jmp end;
+        call mul;  // 0x2
+        jmp end;
+        call sub;  // 0x3
+        jmp end;
+        call div;  // 0x4
+        jmp end;
+        call sdiv;  // 0x5
+        jmp end;
+        call mod;  // 0x6
+        jmp end;
+        call smod;  // 0x7
+        jmp end;
+        call addmod;  // 0x8
+        jmp end;
+        call mulmod;  // 0x9
+        jmp end;
+        call exp;  // 0xa
+        jmp end;
+        call signextend;  // 0xb
+        jmp end;
+
+        end:
+        // Parse results from call
+        let range_check_ptr = [ap - 3];
+        tempvar result = new Uint256([ap - 2], [ap - 1]);
+
+        // Retrieve stack from locals
+        let stack = cast([fp], model.Stack*);
+
+        // Rebind function args with fp
+        let syscall_ptr = cast([fp - 8], felt*);
+        let pedersen_ptr = cast([fp - 7], HashBuiltin*);
+        let bitwise_ptr = cast([fp - 5], BitwiseBuiltin*);
+        let ctx = cast([fp - 4], model.ExecutionContext*);
+        let opcode = cast([fp - 3], model.Opcode*);
+
+        // Finalize opcode
+        let stack = Stack.push(stack, result);
+        let ctx = ExecutionContext.update_stack(ctx, stack);
+        let ctx = ExecutionContext.increment_gas_used(self=ctx, inc_value=opcode.gas);
+        return ctx;
+    }
+
+    func add{range_check_ptr}(popped: Uint256*) -> Uint256 {
```

Comment:
is there an argument for using Uint256 or Uint256* in the return signature?

Reply:
This is the starkware’s default, I’m just keeping the same. Furthermore a pointer cost one more (the size of the pointer)

Reply:
eventually, in the main body I had to create a pointer with `new` otherwise the VM was raising "unsafe pointer used" when doing 

```
let result = cast([ap - 2], Uint256*);
```

instead of

```
tempvar result = new Uint256([ap - 2], [ap - 1]);
```

Reply:
ok!